_posts/2016-05-25-coordination-in-akka.md

```diff
@@ -0,0 +1,147 @@
+
+## Lifecycle
+
+Actors, threads, objects, resources... All of them have different states during theirs lifetime. Some of those states are internal and they shape the entity's behaviour when receiving an external stimulus. However there are state transitions that should be handled by external entities: creation, error recovery and disposal.
+
+## Actor Pools
+
+A long time ago dependency injection showed us that an object should not be in charge of creating its dependencies. Externalising those concerns into a central factory makes code more testable, reusable and readable. We'll need different creation strategies disregarding whether we use DI frameworks or poor man's dependency injection. [Guice Scopes](https://github.com/google/guice/wiki/Scopes) is a good example about how to create one instance per application (@Singleton) or one per scope (@SessionScoped and @RequestScoped).
+
+Often the resources are scarce, expensive to create or they're under heavy load. That means that we can't afford nor a Singleton neither an unbounded creation strategy like [Prototype in Spring](http://www.tutorialspoint.com/spring/spring_bean_scopes.htm). A good example are [thread pools](https://docs.oracle.com/javase/tutorial/essential/concurrency/pools.html).
+
+> Using worker threads minimizes the overhead due to thread creation. Thread objects use a significant amount of memory, and in a large-scale application, allocating and deallocating many thread objects creates a significant memory management overhead.
+
+[Akka Actors](http://doc.akka.io/docs/akka/current/scala/actors.html) are extremely lightweight. We could afford creating millions of them if we knew that they would be disposed quickly. However that's not the case with our use case:
+
+<img src="/assets/img/custom/blog/fsm.png" alt="FSM architecture" title="FSM architecture" class="img img-center img-responsive style-screengrab">
+
+As you can see Finite State Machine (FSM) actors will wait until the Items service finishes deleting the items. It's important to note that the actor is waiting, but not blocking. Actors are attached to [Dispatchers](http://doc.akka.io/docs/akka/2.4.6/scala/dispatchers.html) that have a Thread Pool inside. Those thread pools have a really limited number of threads; blocking in one of the actors would mean running out of threads pretty quickly. Therefore our intent here is bounding the number of actors in our system, not threads. Maybe is an overkill for this particular example, but the point of this series is mainly educational. Let's see a diagram of our service with the proper level of abstraction.
```

Comment:
Therefore our intent here is bounding the number of actors in our system, not threads. ==> what do you mean here ? limiting the number of threads ?